### ISSUE TITLE
Consider supporting Arrow datasets

### ISSUE THREAD

**Issue by petrbouchal** [krlmlr/dm#463]
I have no idea whether it would be technically feasible, but if so the API should probably follow that of your DB-facing functions.

Here is my rationale:

- My understanding from a limited time trying out {dm} is that a dm object still contains copies of the data frames, i.e. they have to all be in-memory and {dm} then duplicates them into the new dm object. This can get unweildy for large-ish data sitting on disk in files or generated in the current process, as we need to have all in memory and creating the dm object doubles the memory footprint. (Though correct me if I am wrong on this.) 
- While it might be useful to first get the data into a DB and then build a dm from the DB, that creates extra friction (and defeats part of the purpose I suspect). 
- {arrow} could provide a happy medium in this workflow: it makes it easy to create partitioned datasets on disk, without necessarily specifying a schema etc., that are fast to query and subset; the dplyr interface to arrow datasets is, from the user's point of view, similar to dplyr's interface to DBI connections. When working with large-ish data that do not merit building a DB, this is my preferred storage/querying approach; it might become more widely used as arrow matures.

...so being able to build a data model from several arrow datasets would be very useful.

The basics of arrow datasets are described in of https://arrow.apache.org/docs/r/articles/dataset.html.

**Comment by krlmlr** [comment]
@nealrichardson: This *might* work out of the box if we relax a few internal checks, and would enable very powerful operations with `dm_wrap_tbl()` and friends if we had an interface for `nest()` and `unnest()` . What's a good way to move forward?

**Comment by krlmlr** [comment]
Thanks.

> ... `tidyr::nest()` and `tidyr::unnest()` cannot be implement before conversion to `List<Struct>` will be accessible.

Perhaps the implementation of `nest()` is a bit simpler if the data is already ordered by the grouping (?) columns. The implementation of `unnest()` might be a bit easier.

For conversion to `List<Struct>`, I have found https://issues.apache.org/jira/browse/ARROW-8779?jql=text%20~%20%22conversion%20list%3Cstruct%3E%22 which seems resolved.

**Comment by krlmlr** [comment]
...but perhaps it's not resolved yet, or won't help nesting. The linked PR seems to discuss creation of nested structures from R, but for efficient `nest()` it might be best if Arrow handled the entire computation?

**Comment by nealrichardson** [comment]
Right, IIUC nest is essentially group_by and aggregate, where the aggregation function is essentially make_struct_array or something similar. That isn't supported as an aggregation function in arrow, and we don't have user-defined aggregation functions yet either.
